Alpha35: Empty event expression produce cryptic error

Related: #3664 (should be fixed together)


I have a template with the following element:
```html
<input type="file" id="picture" accept="image/*"
                           (change)="onFileSelected(f, $event)"
                           (fileread)>
```
On alpha 33 it worked great.
On alpha35 it produces the following runtime error:
```
EXCEPTION: TypeError: Cannot set property 'lastInBinding' of undefined
angular2.js:13030 STACKTRACE:
angular2.js:13030 TypeError: Cannot set property 'lastInBinding' of undefined
    at Function.execute._ConvertAstIntoProtoRecords.create (http://localhost:5555/lib/angular2.js:18104:47)
    at http://localhost:5555/lib/angular2.js:17730:49
    at Array.map (native)
    at createEventRecords (http://localhost:5555/lib/angular2.js:17729:36)
    at JitProtoChangeDetector.execute.JitProtoChangeDetector._createFactory (http://localhost:5555/lib/angular2.js:17575:39)
    at new JitProtoChangeDetector (http://localhost:5555/lib/angular2.js:17567:32)
    at $__1.execute.JitChangeDetection.createProtoChangeDetector (http://localhost:5555/lib/angular2.js:10323:20)
    at http://localhost:5555/lib/angular2.js:15553:42
    at Array.map (native)
    at Function.execute.ListWrapper.map (http://localhost:5555/lib/angular2.js:6374:26)
```

The following modification still produces the error:
```html
<input type="file" id="picture" accept="image/*"
                           (change)="onFileSelected(f, $event)"
                           (fileread)="">
```

This one works
```html
<input type="file" id="picture" accept="image/*"
                           (change)="onFileSelected(f, $event)"
                           (fileread)="x">
```

This error is still occuring in RC1 when using something like `<a (click)></a>)`
```
browser_adapter.ts:88 EXCEPTION: TypeError: Cannot read property 'visitStatement' of undefined
2016-05-07 07:15:03.983 browser_adapter.ts:78 EXCEPTION: TypeError: Cannot read property 'visitStatement' of undefinedBrowserDomAdapter.logError @ browser_adapter.ts:78BrowserDomAdapter.logGroup @ browser_adapter.ts:89ExceptionHandler.call @ exception_handler.ts:53(anonymous function) @ application_ref.ts:341ZoneDelegate.invoke @ zone.js:323NgZoneImpl.inner.inner.fork.onInvoke @ ng_zone_impl.ts:67ZoneDelegate.invoke @ zone.js:322Zone.run @ zone.js:216(anonymous function) @ zone.js:571ZoneDelegate.invokeTask @ zone.js:356NgZoneImpl.inner.inner.fork.onInvokeTask @ ng_zone_impl.ts:56ZoneDelegate.invokeTask @ zone.js:355Zone.runTask @ zone.js:256drainMicroTaskQueue @ zone.js:474ZoneTask.invoke @ zone.js:426
2016-05-07 07:15:03.985 browser_adapter.ts:88 EXCEPTION: Error: Uncaught (in promise): TypeError: Cannot read property 'visitStatement' of undefined
2016-05-07 07:15:03.985 browser_adapter.ts:78 EXCEPTION: Error: Uncaught (in promise): TypeError: Cannot read property 'visitStatement' of undefinedBrowserDomAdapter.logError @ browser_adapter.ts:78BrowserDomAdapter.logGroup @ browser_adapter.ts:89ExceptionHandler.call @ exception_handler.ts:53(anonymous function) @ application_ref.ts:304schedulerFn @ async.ts:131SafeSubscriber.__tryOrUnsub @ Subscriber.ts:240SafeSubscriber.next @ Subscriber.ts:192Subscriber._next @ Subscriber.ts:133Subscriber.next @ Subscriber.ts:93Subject._finalNext @ Subject.ts:154Subject._next @ Subject.ts:144Subject.next @ Subject.ts:90EventEmitter.emit @ async.ts:117NgZone._zoneImpl.ng_zone_impl_1.NgZoneImpl.onError @ ng_zone.ts:138NgZoneImpl.inner.inner.fork.onHandleError @ ng_zone_impl.ts:90ZoneDelegate.handleError @ zone.js:327Zone.runGuarded @ zone.js:233_loop_1 @ zone.js:487drainMicroTaskQueue @ zone.js:494ZoneTask.invoke @ zone.js:426
2016-05-07 07:15:03.985 browser_adapter.ts:78 STACKTRACE:BrowserDomAdapter.logError @ browser_adapter.ts:78ExceptionHandler.call @ exception_handler.ts:56(anonymous function) @ application_ref.ts:304schedulerFn @ async.ts:131SafeSubscriber.__tryOrUnsub @ Subscriber.ts:240SafeSubscriber.next @ Subscriber.ts:192Subscriber._next @ Subscriber.ts:133Subscriber.next @ Subscriber.ts:93Subject._finalNext @ Subject.ts:154Subject._next @ Subject.ts:144Subject.next @ Subject.ts:90EventEmitter.emit @ async.ts:117NgZone._zoneImpl.ng_zone_impl_1.NgZoneImpl.onError @ ng_zone.ts:138NgZoneImpl.inner.inner.fork.onHandleError @ ng_zone_impl.ts:90ZoneDelegate.handleError @ zone.js:327Zone.runGuarded @ zone.js:233_loop_1 @ zone.js:487drainMicroTaskQueue @ zone.js:494ZoneTask.invoke @ zone.js:426
2016-05-07 07:15:03.985 browser_adapter.ts:78 Error: Uncaught (in promise): TypeError: Cannot read property 'visitStatement' of undefined
    at resolvePromise (zone.js:538)
    at PromiseCompleter.reject (zone.js:515)
    at eval (application_ref.ts:340)
    at ZoneDelegate.invoke (zone.js:323)
    at Object.NgZoneImpl.inner.inner.fork.onInvoke (ng_zone_impl.ts:67)
    at ZoneDelegate.invoke (zone.js:322)
    at Zone.run (zone.js:216)
    at zone.js:571
    at ZoneDelegate.invokeTask (zone.js:356)
    at Object.NgZoneImpl.inner.inner.fork.onInvokeTask (ng_zone_impl.ts:56)BrowserDomAdapter.logError @ browser_adapter.ts:78ExceptionHandler.call @ exception_handler.ts:57(anonymous function) @ application_ref.ts:304schedulerFn @ async.ts:131SafeSubscriber.__tryOrUnsub @ Subscriber.ts:240SafeSubscriber.next @ Subscriber.ts:192Subscriber._next @ Subscriber.ts:133Subscriber.next @ Subscriber.ts:93Subject._finalNext @ Subject.ts:154Subject._next @ Subject.ts:144Subject.next @ Subject.ts:90EventEmitter.emit @ async.ts:117NgZone._zoneImpl.ng_zone_impl_1.NgZoneImpl.onError @ ng_zone.ts:138NgZoneImpl.inner.inner.fork.onHandleError @ ng_zone_impl.ts:90ZoneDelegate.handleError @ zone.js:327Zone.runGuarded @ zone.js:233_loop_1 @ zone.js:487drainMicroTaskQueue @ zone.js:494ZoneTask.invoke @ zone.js:426
2016-05-07 07:15:03.987 zone.js:461 Unhandled Promise rejection: Cannot read property 'visitStatement' of undefined ; Zone: angular ; Task: Promise.then ; Value: TypeError: Cannot read property 'visitStatement' of undefined(…)consoleError @ zone.js:461_loop_1 @ zone.js:490drainMicroTaskQueue @ zone.js:494ZoneTask.invoke @ zone.js:426
2016-05-07 07:15:03.989 zone.js:463 Error: Uncaught (in promise): TypeError: Cannot read property 'visitStatement' of undefined(…)consoleError @ zone.js:463_loop_1 @ zone.js:490drainMicroTaskQueue @ zone.js:494ZoneTask.invoke @ zone.js:426
```

--- a/modules/@angular/compiler/src/expression_parser/parser.ts
+++ b/modules/@angular/compiler/src/expression_parser/parser.ts
@@ -43,8 +43,8 @@ export class Parser {
       input: string, location: any,
       interpolationConfig: InterpolationConfig = DEFAULT_INTERPOLATION_CONFIG): ASTWithSource {
     this._checkNoInterpolation(input, location, interpolationConfig);
-    var tokens = this._lexer.tokenize(this._stripComments(input));
-    var ast = new _ParseAST(input, location, tokens, true, this.errors).parseChain();
+    const tokens = this._lexer.tokenize(this._stripComments(input));
+    const ast = new _ParseAST(input, location, tokens, true, this.errors).parseChain();
     return new ASTWithSource(ast, input, location, this.errors);
   }
 
@@ -155,15 +155,15 @@ export class Parser {
   }
 
   private _stripComments(input: string): string {
-    let i = this._commentStart(input);
+    const i = this._commentStart(input);
     return isPresent(i) ? input.substring(0, i).trim() : input;
   }
 
   private _commentStart(input: string): number {
     var outerQuote: number = null;
-    for (var i = 0; i < input.length - 1; i++) {
-      let char = StringWrapper.charCodeAt(input, i);
-      let nextChar = StringWrapper.charCodeAt(input, i + 1);
+    for (let i = 0; i < input.length - 1; i++) {
+      const char = StringWrapper.charCodeAt(input, i);
+      const nextChar = StringWrapper.charCodeAt(input, i + 1);
 
       if (char === chars.$SLASH && nextChar == chars.$SLASH && isBlank(outerQuote)) return i;
 

--- a/modules/@angular/compiler/src/template_parser/template_parser.ts
+++ b/modules/@angular/compiler/src/template_parser/template_parser.ts
@@ -12,7 +12,7 @@ import {Console, MAX_INTERPOLATION_VALUES} from '../../core_private';
 import {ListWrapper, StringMapWrapper, SetWrapper,} from '../facade/collection';
 import {RegExpWrapper, isPresent, isBlank} from '../facade/lang';
 import {BaseException} from '../facade/exceptions';
-import {AST, Interpolation, ASTWithSource, TemplateBinding, RecursiveAstVisitor, BindingPipe, ParserError} from '../expression_parser/ast';
+import {EmptyExpr, AST, Interpolation, ASTWithSource, TemplateBinding, RecursiveAstVisitor, BindingPipe, ParserError} from '../expression_parser/ast';
 import {Parser} from '../expression_parser/parser';
 import {CompileDirectiveMetadata, CompilePipeMetadata, CompileTokenMetadata, removeIdentifierDuplicates,} from '../compile_metadata';
 import {HtmlParser, ParseTreeResult} from '../html_parser/html_parser';
@@ -209,7 +209,7 @@ class TemplateParseVisitor implements html.Visitor {
   }
 
   private _reportParserErors(errors: ParserError[], sourceSpan: ParseSourceSpan) {
-    for (let error of errors) {
+    for (const error of errors) {
       this._reportError(error.message, sourceSpan);
     }
   }
@@ -234,9 +234,16 @@ class TemplateParseVisitor implements html.Visitor {
 
   private _parseAction(value: string, sourceSpan: ParseSourceSpan): ASTWithSource {
     const sourceInfo = sourceSpan.start.toString();
+
     try {
       const ast = this._exprParser.parseAction(value, sourceInfo, this._interpolationConfig);
-      if (ast) this._reportParserErors(ast.errors, sourceSpan);
+      if (ast) {
+        this._reportParserErors(ast.errors, sourceSpan);
+      }
+      if (!ast || ast.ast instanceof EmptyExpr) {
+        this._reportError(`Empty expressions are not allowed`, sourceSpan);
+        return this._exprParser.wrapLiteralPrimitive('ERROR', sourceInfo);
+      }
       this._checkPipes(ast, sourceSpan);
       return ast;
     } catch (e) {

--- a/modules/@angular/compiler/test/template_parser/template_parser_spec.ts
+++ b/modules/@angular/compiler/test/template_parser/template_parser_spec.ts
@@ -328,8 +328,18 @@ export function main() {
       describe('events', () => {
 
         it('should parse bound events with a target', () => {
-          expect(humanizeTplAst(parse('<div (window:event)="v">', [
-          ]))).toEqual([[ElementAst, 'div'], [BoundEventAst, 'event', 'window', 'v']]);
+          expect(humanizeTplAst(parse('<div (window:event)="v">', []))).toEqual([
+            [ElementAst, 'div'],
+            [BoundEventAst, 'event', 'window', 'v'],
+          ]);
+        });
+
+        it('should report an error on empty expression', () => {
+          expect(() => parse('<div (event)="">', []))
+              .toThrowError(/Empty expressions are not allowed/);
+
+          expect(() => parse('<div (event)="  ">', []))
+              .toThrowError(/Empty expressions are not allowed/);
         });
 
         it('should parse bound events via (...) and not report them as attributes', () => {
@@ -519,10 +529,10 @@ export function main() {
       });
 
       describe('providers', () => {
-        var nextProviderId: any /** TODO #9100 */;
+        var nextProviderId: number;
 
         function createToken(value: string): CompileTokenMetadata {
-          var token: any /** TODO #9100 */;
+          let token: CompileTokenMetadata;
           if (value.startsWith('type:')) {
             token = new CompileTokenMetadata({
               identifier:
@@ -555,8 +565,8 @@ export function main() {
         }
 
         function createProvider(
-            token: string, {multi = false, deps = /*@ts2dart_const*/[]}:
-                               {multi?: boolean, deps?: string[]} = {}): CompileProviderMetadata {
+            token: string, {multi = false, deps = []}: {multi?: boolean, deps?: string[]} = {}):
+            CompileProviderMetadata {
           return new CompileProviderMetadata({
             token: createToken(token),
             multi: multi,
@@ -566,8 +576,7 @@ export function main() {
         }
 
         function createDir(
-            selector: string, {providers = null, viewProviders = null, deps = /*@ts2dart_const*/[],
-                               queries = /*@ts2dart_const*/[]}: {
+            selector: string, {providers = null, viewProviders = null, deps = [], queries = []}: {
               providers?: CompileProviderMetadata[],
               viewProviders?: CompileProviderMetadata[],
               deps?: string[],

